File: 2296.design-a-text-editor.cpp
#
# @lc app=leetcode id=2296 lang=cpp
#
# [2296] Design a Text Editor
#

# @lc code=start
class TextEditor {
public:
    TextEditor() {}

    void addText(string text) {
        for (char c : text) {
            left.push_back(c);
        }
    }

    int deleteText(int k) {
        int deleted = 0;
        for (int i = 0; i < k; ++i) {
            if (left.empty()) {
                break;
            }
            left.pop_back();
            ++deleted;
        }
        return deleted;
    }

    string cursorLeft(int k) {
        int steps = std::min(k, (int)left.size());
        for (int i = 0; i < steps; ++i) {
            right.push_front(left.back());
            left.pop_back();
        }
        return getScreen();
    }

    string cursorRight(int k) {
        int steps = std::min(k, (int)right.size());
        for (int i = 0; i < steps; ++i) {
            left.push_back(right.front());
            right.pop_front();
        }
        return getScreen();
    }

private:
    std::deque<char> left;
    std::deque<char> right;

    std::string getScreen() const {
        size_t n = left.size();
        if (n == 0) return "";
        size_t start = (n > 10 ? n - 10 : 0);
        std::string screen;
        screen.reserve(10);
        for (size_t i = start; i < n; ++i) {
            screen += left[i];
        }
        return screen;
    }
};

/**
 * Your TextEditor object will be instantiated and called as such:
 * TextEditor* obj = new TextEditor();
 * obj->addText(text);
 * int param_2 = obj->deleteText(k);
 * string param_3 = obj->cursorLeft(k);
 * string param_4 = obj->cursorRight(k);
 */
# @lc code=end

File: 2296.design-a-text-editor.py
#
# @lc app=leetcode id=2296 lang=python3
#
# [2296] Design a Text Editor
#

# @lc code=start
from collections import deque

class TextEditor:

    def __init__(self):
        self.left = []
        self.right = deque()

    def addText(self, text: str) -> None:
        for c in text:
            self.left.append(c)

    def deleteText(self, k: int) -> int:
        deleted = min(k, len(self.left))
        for _ in range(deleted):
            self.left.pop()
        return deleted

    def cursorLeft(self, k: int) -> str:
        moves = min(k, len(self.left))
        for _ in range(moves):
            self.right.appendleft(self.left.pop())
        return ''.join(self.left[-10:])

    def cursorRight(self, k: int) -> str:
        moves = min(k, len(self.right))
        for _ in range(moves):
            self.left.append(self.right.popleft())
        return ''.join(self.left[-10:])


# Your TextEditor object will be instantiated and called as such:
# obj = TextEditor()
# obj.addText(text)
# param_2 = obj.deleteText(k)
# param_3 = obj.cursorLeft(k)
# param_4 = obj.cursorRight(k)
# @lc code=end